cmake: Enable CMAKE_EXPORT_COMPILE_COMMANDS

Pennycook opened the pull request:
Enabling this option causes CMake to generate a compile_commands.json file containing a compilation database that can be used to drive third-party tools.

Pennycook committed:
cmake: Enable CMAKE_EXPORT_COMPILE_COMMANDS

Enabling this option causes CMake to generate a compile_commands.json file
containing a compilation database that can be used to drive third-party tools.

haampie commented:
Should go under `_conditional_cmake_defaults` as it only exists in `cmake@3.5:`.

Also not 100% sure if there are any downsides to doing this. Do some projects have special logic that influence where/how this json file is generated or is this the only on/off toggle?

Pennycook committed:
cmake: Make compile commands conditional

CMAKE_EXPORT_COMPILE_COMMANDS only exists for CMake >= 3.5

Pennycook commented:
> Should go under `_conditional_cmake_defaults` as it only exists in `cmake@3.5:`.

Ah, I didn't realize that.  Fixed in 245b1553df0aae904387b6726f8b3f9e91032e2a.

> Also not 100% sure if there are any downsides to doing this. Do some projects have special logic that influence where/how this json file is generated or is this the only on/off toggle?

I believe that this is the only on/off toggle, and the compile_commands.json file always ends up in the build directory.

I'm not aware of any downsides to generating a compilation database unconditionally.  I did consider an alternative design, adding a `--compile-commands` option to `install` similar to the `--source` option, but I couldn't figure out how to get the information into all the right places.

One potential issue is that `compile_commands.json` won't necessarily be generated for all CMake packages. It needs CMake >= 3.5 (as you pointed out) and only works for Unix makefiles and Ninja.  What does spack do if there's a file listed in `archive_files` and it doesn't exist?

becker33 commented on lib/spack/spack/build_systems/cmake.py:
```diff
@@ -284,7 +288,10 @@ class CMakeBuilder(BaseBuilder):
     def archive_files(self):
         """Files to archive for packages based on CMake"""
-        return [os.path.join(self.build_directory, "CMakeCache.txt")]
+        return [
+            os.path.join(self.build_directory, "CMakeCache.txt"),
+            os.path.join(self.build_directory, "compile_commands.json"),
+        ]
```
```suggestion
        files = [
            os.path.join(self.build_directory, "CMakeCache.txt"),
        ]
        
        if cmake.satisfies("@3.5:"):
            os.path.join(self.build_directory, "compile_commands.json"),
```
And this could probably have `if cmake.satisfies(compilation_database_constraints)` to avoid duplicating code with line 99

Pennycook replied:
Good idea. I couldn't find a way to describe the requirements as a single spec string, so I added a `_supports_compilation_databases` function in https://github.com/spack/spack/pull/42353/commits/bfda1626b8a8c94586c3376f6392b3979847db8e and used it in both places.

becker33 commented on lib/spack/spack/build_systems/cmake.py:
```diff
@@ -95,6 +95,10 @@ def _conditional_cmake_defaults(pkg: spack.package_base.PackageBase, args: List[
         args.append(CMakeBuilder.define("CMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY", False))
 
+    # CMAKE_EXPORT_COMPILE_COMMANDS only exists for CMake >= 3.5
+    if cmake.satisfies("@3.5:"):
```
Can we add the rest of the constraints here? We should only set this under the conditions it will actually work, so we can reuse the constraints below and avoid setting it in `archive_files` when it will fail.

Pennycook replied:
The new `_supports_compilation_databases` function from https://github.com/spack/spack/pull/42353/commits/bfda1626b8a8c94586c3376f6392b3979847db8e now checks the CMake version and the generator.  These are the only restrictions I'm aware of.

The `CMAKE_EXPORT_COMPILE_COMMANDS` documentation says that the feature "doesn't work well" with `CMAKE_UNITY_BUILD`, but I'm honestly not sure how to interpret that!

Pennycook committed:
cmake: Add extra conditions for compile commands

Exporting compilation databases is only supported for Makefile and Ninja
generators, so check these conditions as well.
